mcs/class/System/Mono.AppleTls/AppleTlsContext.cs

```diff
@@ -0,0 +1,931 @@
+				if (context.lastException == null)
+					context.lastException = ex;
+				return SslStatus.Internal;
+			}
+		}
+
+		SslStatus NativeReadCallback (IntPtr data, ref /*HACK*/ IntPtr dataLength)
+		{
+			if (closed || disposed || Parent == null)
+				return SslStatus.ClosedAbort;
+
+			var len = (int)dataLength;
+			var readBuffer = new byte [len];
+
+			Debug ("NativeReadCallback: {0} {1}", dataLength, len);
+
+			bool wantMore;
+			var ret = Parent.InternalRead (readBuffer, 0, len, out wantMore);
+			dataLength = (IntPtr)ret; // HACK
+
+			Debug ("NativeReadCallback #1: {0} - {1} {2}", len, ret, wantMore);
+
+			if (ret < 0)
+				return SslStatus.ClosedAbort;
+
+			Marshal.Copy (readBuffer, 0, data, ret);
+
+			if (ret > 0)
+				return SslStatus.Success;
+			else if (wantMore)
+				return SslStatus.WouldBlock;
+			else if (ret == 0) {
+				closedGraceful = true;
+				return SslStatus.ClosedGraceful;
+			} else {
+				return SslStatus.Success;
+			}
+		}
+
+		SslStatus NativeWriteCallback (IntPtr data, ref /*HACK*/ IntPtr dataLength)
+		{
+			if (closed || disposed || Parent == null)
+				return SslStatus.ClosedAbort;
+
+			var len = (int)dataLength;
+			var writeBuffer = new byte [len];
+
+			Marshal.Copy (data, writeBuffer, 0, len);
+
+			Debug ("NativeWriteCallback: {0}", len);
+
+			var ok = Parent.InternalWrite (writeBuffer, 0, len);
+
+			Debug ("NativeWriteCallback done: {0} {1}", len, ok);
+
+			return ok ? SslStatus.Success : SslStatus.ClosedAbort;
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLRead (/* SSLContextRef */ IntPtr context, /* const void* */ byte* data, /* size_t */ nint dataLength, /* size_t* */ out nint processed);
+
+		public override unsafe int Read (byte[] buffer, int offset, int count, out bool wantMore)
+		{
+			if (Interlocked.Exchange (ref pendingIO, 1) == 1)
+				throw new InvalidOperationException ();
+
+			Debug ("Read: {0},{1}", offset, count);
+
+			lastException = null;
+
+			try {
+				nint processed;
+				SslStatus status;
+
+				fixed (byte *d = &buffer [offset])
+					status = SSLRead (Handle, d, (IntPtr)count, out processed);
+
+				Debug ("Read done: {0} {1} {2}", status, count, processed);
+
+				if (closedGraceful && (status == SslStatus.ClosedAbort || status == SslStatus.ClosedGraceful)) {
+					/*
+					 * This is really ugly, but unfortunately SSLRead() also returns 'SslStatus.ClosedAbort'
+					 * when the first inner Read() returns 0.  MobileAuthenticatedStream.InnerRead() attempts
+					 * to distinguish between a graceful close and abnormal termination of connection.
+					 */
+					wantMore = false;
+					return 0;
+				}
+
+				CheckStatusAndThrow (status, SslStatus.WouldBlock, SslStatus.ClosedGraceful);
+				wantMore = status == SslStatus.WouldBlock;
+				return (int)processed;
+			} catch (Exception ex) {
+				Debug ("Read error: {0}", ex);
+				throw;
+			} finally {
+				pendingIO = 0;
```

Comment:
@marek-safar I do not feel comfortable making non-mechanical changes to this code, so i should be @baulig .

Do you want it fixed before landing this?